### ISSUE TITLE
Check performance of rust checks

### ISSUE THREAD

**Issue by rmcgibbo** [jtojnar/nixpkgs-hammering#79]
I think I'm currently seeing a case where each of the rust AST checks took about 2.5 minutes to run, at 100% CPU. If I understand my logs correctly, I thought they were only processing 216 packages. If I don't understand correctly, it might be 8,000 packages. Either way, it seems a bit excessive.

I will try to reproduce/investigate in a more controlled environment.

**Comment by rmcgibbo** [comment]
Unfortunately, I think this is real:

```
echo '[{"name": "haskellPackages.profunctors", "location": {"file": "/home/mcgibbon/projects/nixpkgs/pkgs/development/haskell-modules/hackage-packages.nix", "line": 208189, "column": null}, "drv": null, "output": null}]' | time cargo run --bin missing-patch-comment --release
   Compiling nixpkgs-hammering-ast-checks v0.0.0 (/home/mcgibbon/projects/nixpkgs-hammering/rust-checks)
    Finished release [optimized] target(s) in 2.60s
     Running `target/release/missing-patch-comment`
{"haskellPackages.profunctors":[]}
9.32user 0.23system 0:04.38elapsed 217%CPU (0avgtext+0avgdata 226420maxresident)k
0inputs+4848outputs (0major+71828minor)pagefaults 0swaps
```

Looks like running `missing-patch-comment` on a single attr in haskel-packages.nix takes 9 seconds on my machine. With a few hundred packages, many of which could have all been in the same file causing it to be analyzed repeatedly, the initial observation was probably legitimate.

**Comment by jtojnar** [comment]
I will try to convert the check to the tree traversal as described in https://github.com/jtojnar/nixpkgs-hammering/pull/32#issuecomment-776936922 in a few weeks, hopefully it will be more efficient.

But I guess we should ignore generated files either way.

**Comment by rmcgibbo** [comment]
I don't think that will help. I measured just [parsing the file](https://github.com/jtojnar/nixpkgs-hammering/blob/master/rust-checks/src/analysis.rs#L61-L71) to take 1.3 seconds.

I think the best options are:
1. Amortize the parser costs. Currently if run hammering on multiple attrs from haskellPackages, the entire file is parsed once for each attr for each check. It would be pretty easy to re-arrange [analyze_nix_file](https://github.com/jtojnar/nixpkgs-hammering/blob/master/rust-checks/src/analysis.rs#L14) a bit so to that each file was only parsed once per check. It would also be possible, but more annoying, to merge all of the AST checks into a single binary rather than having one binary per rule. THis would let us parse each nix file only once over all of the checks.
2. Ignore `hackage-packages.nix` / `haskellPackages`.
